### PR TITLE
Start PetscObject

### PR DESCRIPTION
**Pull request by ZoeLeibowitz**
Playing around with PR UI.

### CODE REVIEW

**Comment by edsml-zl5621** [comment]
Start of PetscObject ... please review if you have time, thanks!

**Comment by mloubout** [comment]
@edsml-zl5621 I think can make use of the standard devito objects and interface rather than resulting in all these new classe. Below is an example of something that would work as a nice extension without needing to go all the way down to `object` but might need some tweaks

```
from ctypes import POINTER

import numpy as np

from devito import *

from devito.ir.iet import Definition, DummyExpr
from devito.types.dense import DiscreteFunction
from devito.types.basic import Basic, AbstractSymbol
from devito.types.object import AbstractObject
from devito.tools import dtype_to_cstr, dtype_to_ctype


class PetscObject(AbstractSymbol):
    is_PetscObject = True
    """
    PetscObjects encode their dtype as a class attribute.
    """

    @property
    def is_const(self):
        return self._is_const
    
    @property
    def dtype(self):
        return self._dtype

    @property
    def _C_name(self):
        return self.name

    @property
    def _C_ctype(self):
        name = dtype_to_cstr(self._dtype).capitalize()
        return type('Petsc%s' % name, (dtype_to_ctype(self._dtype),), {})


class PetscFunction(Function):

    @property
    def _C_ctype(self):
        ctypename = 'Petsc%s' % dtype_to_cstr(self.dtype).capitalize()
        ctype = dtype_to_ctype(self.dtype)
        r = POINTER(type(ctypename, (ctype,), {}))
        for n in range(len(self.dimensions)-1):
            r = POINTER(r)
        return r


grid = Grid((2, 2))
x, y = grid.dimensions
# pointer and const functionality
ptr1 = PetscFunction(name='ptr1', dtype=np.int32, grid=grid, dimensions=(x,), shape=(2,), is_const=True)
ptr2 = PetscFunction(name='ptr2', dtype=np.int32, grid=grid, dimensions=(x,), shape=(2,))
pptr = PetscFunction(name='pptr', dtype=np.int32, grid=grid)

defn1 = Definition(ptr1)
defn2 = Definition(ptr2)
defn3 = Definition(pptr)

# algebraic expressions
x = PetscObject(name='x', dtype=np.int32)
y = PetscObject(name='y', dtype=np.int32)

expr1 = DummyExpr(x, 15, init=True)
expr2 = DummyExpr(x, y-1)

print(str(defn1))
print(str(defn2))
print(str(defn3))
print(str(expr1))
print(str(expr2))

assert str(defn1) == "const PetscInt * ptr1;"
assert str(defn2) == "PetscInt * ptr2;"
assert str(defn3) == "PetscInt ** pptr;"
assert str(expr1) == "PetscInt x = 15;"
assert str(expr2) == "x = -1 + y;"
```

File: devito/types/object.py [hunk 1]
```diff
@@ -220,3 +341,87 @@ def _mem_internal_eager(self):
...
+
+    def __init__(self, name, dtype, **kwargs):
+        self.name = name
+        self._dtype = dtype
+        self._is_const = kwargs.get('is_const', False)
+
+    # taken exactly from Function class but removed the "shape_global"
+    #  # and also added the option for all dimensions, shape and grid to be None
+    @classmethod
+    def __shape_setup__(cls, **kwargs):
```
Comment (FabioLuporini) [inline devito/types/object.py:358]: both shape_setup and indices_setup should belong to AbstractObjectWithShape

File: devito/types/object.py [hunk 2]
```diff
@@ -220,3 +341,87 @@ def _mem_internal_eager(self):
...
+            raise TypeError("`dimensions` required if both `grid` and "
+                            "`shape` are provided")
+        else:
+            # Got `grid`, `dimensions`, and `shape`. We sanity-check that the
+            # Dimensions in `dimensions` also appearing in `grid` have same size
+            # (given by `shape`) as that provided in `grid`
+            if len(shape) != len(dimensions):
+                raise ValueError("`shape` and `dimensions` must have the "
+                                 "same number of entries")
+            loc_shape = []
```
Comment (FabioLuporini) [inline devito/types/object.py:383]: no distinction between global and local shape (MPI irrelevant for Objects), so we don't need this part

File: devito/types/object.py [hunk 3]
```diff
@@ -84,6 +87,124 @@ def function(self):
...
+    Base class for objects with derived type.
+
+    The hierarchy is structured as follows
+
+                         AbstractObjectWithShape
+                                   |
+                              PetscObject
+    """
+
+    AbstractObjectWithShape = True
```
Comment (FabioLuporini) [inline devito/types/object.py:102]: don't add unless proven useful. Too early

File: devito/types/object.py [hunk 4]
```diff
@@ -84,6 +87,124 @@ def function(self):
...
+class AbstractObjectWithShape(Basic, sympy.Basic, Pickable):
+
+    """
+    Base class for objects with derived type.
+
+    The hierarchy is structured as follows
+
+                         AbstractObjectWithShape
+                                   |
+                              PetscObject
```
Comment (FabioLuporini) [inline devito/types/object.py:99]: a PetscObject is way too special to live inside the `types/object` directory. Only the more general class belong here. 

The PetscObject type should instead be defined wherever you will implement the petsc machinery (`passes/iet/petsc/` ?) . Just like we do for some MPI Objects under `/mpi/distributed` or `/mpi/routines`

File: devito/types/object.py (hunk 1 above)
Comment (ZoeLeibowitz) [reply devito/types/object.py:358]: Does this mean I should remove the 'empty' shape_setup and indices_setup that are currently inside AbstractObjectWithShape and replace them with these ones?

Comment (FabioLuporini) [reply devito/types/object.py:358]: Why not? 

Say I create a new subtype, let's call it `class MyNewObject(AbstractObjectWithShape)`

I want this new type to share all the key PetscObject mechanisms , don't I?

File: devito/types/object.py [hunk 5]
```diff
@@ -1,14 +1,17 @@
 from ctypes import byref
+from ctypes import POINTER
```
Comment (EdCaunt) [inline devito/types/object.py:2]: Why not `from ctypes import byref, POINTER`?

File: devito/types/object.py [hunk 6]
```diff
@@ -220,3 +341,87 @@ def _mem_internal_eager(self):
...
+
+    is_PetscObject = True
+
+    def __init__(self, name, dtype, **kwargs):
+        self.name = name
+        self._dtype = dtype
+        self._is_const = kwargs.get('is_const', False)
+
+    # taken exactly from Function class but removed the "shape_global"
+    #  # and also added the option for all dimensions, shape and grid to be None
```
Comment (EdCaunt) [inline devito/types/object.py:356]: Extra `#` in the second comment here

File: devito/types/object.py [hunk 7]
```diff
@@ -220,3 +341,87 @@ def _mem_internal_eager(self):
...
+                    loc_shape.append(s)
+            shape = tuple(loc_shape)
+        return shape
+
+    # taken from Function class but removed staggered indices
+    # and also added the option for all dimensions, shape and grid to be None
+    @classmethod
+    def __indices_setup__(cls, **kwargs):
+        grid = kwargs.get('grid')
+        # shape = kwargs.get('shape', None)
```
Comment (EdCaunt) [inline devito/types/object.py:403]: Leftover comment

File: devito/types/object.py [hunk 8]
```diff
@@ -220,3 +341,87 @@ def _mem_internal_eager(self):
...
+            if dimensions is None:
+                raise TypeError("Need either `grid` or `dimensions`")
+        elif dimensions is None:
+            dimensions = grid.dimensions
+
+        return dimensions, dimensions
+
+    @property
+    def _C_ctype(self):
+        """
```
Comment (EdCaunt) [inline devito/types/object.py:420]: Is the empty docstring intentional?

**Comment by ZoeLeibowitz** [comment]
Hi, I added a FunctionPtr class and a test for the symbolic representation of C's function pointers. I'm not sure if this is the right way to do it so if someone could have a look that would be great! Thank you

File: devito/symbolics/extended_sympy.py [hunk 9]
```diff
@@ -398,6 +398,26 @@ def typ(self):
...
         return '(%s)' % self.typ
+    
+
+class FunctionPtr(UnaryOp):
+
+    """
+    Symbolic representation of C's function pointers.
+    """
+
+    _return_typ = ''
```
Comment (FabioLuporini) [inline devito/symbolics/extended_sympy.py:409]: why aren't these instance attributes?

also, you can add an 'e' at the end of 'type' ;)

File: devito/symbolics/extended_sympy.py [hunk 10]
```diff
@@ -398,6 +398,26 @@ def typ(self):
...
+    _parameter_typ = ''
+
+    def __new__(cls, base, **kwargs):
+        obj = super().__new__(cls, base)
+        return obj
+
+    func = Pickable._rebuild
+
+    @property
+    def _op(self):
```
Comment (FabioLuporini) [inline devito/symbolics/extended_sympy.py:419]: shouldn't this just be `__str__` (== `__repr__`)

**Comment by FabioLuporini** [comment]
Also, how about a first tiny tiny PR that only ships `FunctionPointer` and related tests -- see:

* test_symbolics.py
* test_pickling.py

THe first merge into devito master is big milestone :-)

File: devito/symbolics/extended_sympy.py (hunk 9 above)
Comment (ZoeLeibowitz) [reply devito/symbolics/extended_sympy.py:409]: I took inspo from the Cast class https://github.com/devitocodes/devito/blob/0abe3952a792ed9ba5bf9ef904fa2fec04789a79/devito/symbolics/extended_sympy.py#L370C1-L400C33 and _base_typ is not an instance attribute. Is there a particular reason for this?

**Comment by ZoeLeibowitz** [comment]
> Also, how about a first tiny tiny PR that only ships `FunctionPointer` and related tests -- see:
> 
> * test_symbolics.py
> * test_pickling.py
> 
> THe first merge into devito master is big milestone :-)

**Comment by ZoeLeibowitz** [comment]
> > Also, how about a first tiny tiny PR that only ships `FunctionPointer` and related tests -- see:
> > 
> > * test_symbolics.py
> > * test_pickling.py
> > 
> > THe first merge into devito master is big milestone :-)

File: devito/symbolics/extended_sympy.py (hunk 9 above)
Comment (FabioLuporini) [reply devito/symbolics/extended_sympy.py:409]: I think it's a leftover from code that has been refactorer multiple times. You may probably tweak that too !

File: devito/types/object.py [hunk 11]
```diff
@@ -84,6 +85,177 @@ def function(self):
     __reduce_ex__ = Pickable.__reduce_ex__
 
 
+class AbstractObjectWithShape(Basic, sympy.Basic, Pickable):
+
+    # need to check this?
+    # __rkwargs__ = ('name', 'dtype', 'is_const', 'grid', 'shape', 'dimensions')
+
+    def __new__(cls, *args, **kwargs):
```
Comment (mloubout) [inline devito/types/object.py:93]: This still looks like there is a quite large amount of overlap with `AbstractFunction` so I'm not sure why this is needed.

Comment (FabioLuporini) [reply devito/types/object.py:93]: @mloubout AbstractObject doesn't inherit from AbstractFunction, so unless we lift some of the construction logic to the closest common ancestor (as in, `Basic`), this needs to stay here, I think

Comment (mloubout) [reply devito/types/object.py:93]: But this doesn't need to inheriting from `AbstractObject` if there is already an existing type that does all of it. This just redefines `AbstractFunction` with a new calss name and `_const` it should be a 2 liners inheriting from `AbstractFunction`

**Pull request closed without merging**